src/java/main/org/apache/zookeeper/server/quorum/ObserverMaster.java

```diff
@@ -0,0 +1,513 @@
+
+    @Override
+    public ZKDatabase getZKDatabase() {
+        return zks.getZKDatabase();
+    }
+
+    @Override
+    public void waitForNewLeaderAck(long sid, long zxid) throws InterruptedException {
+        // no need to wait since we are a follower
+    }
+
+    @Override
+    public int getCurrentTick() {
+        return self.tick.get();
+    }
+
+    @Override
+    public void processAck(long sid, long zxid, SocketAddress localSocketAddress) {
+        if ((zxid & 0xffffffffL) == 0) {
+            /*
+             * We no longer process NEWLEADER ack by this method. However,
+             * the learner sends ack back to the leader after it gets UPTODATE
+             * so we just ignore the message.
+             */
+            return;
+        }
+
+        throw new RuntimeException("Observers shouldn't send ACKS ack = " + Long.toHexString(zxid));
+    }
+
+    @Override
+    public void touch(long sess, int to) {
+        zks.getSessionTracker().touchSession(sess, to);
+    }
+
+    boolean revalidateLearnerSession(QuorumPacket qp) throws IOException {
+        ByteArrayInputStream bis = new ByteArrayInputStream(qp.getData());
+        DataInputStream dis = new DataInputStream(bis);
+        long id = dis.readLong();
+        boolean valid = dis.readBoolean();
+        Iterator<Revalidation> itr = pendingRevalidations.iterator();
+        if (!itr.hasNext()) {
+            // not a learner session, handle locally
+            return false;
+        }
+        Revalidation revalidation = itr.next();
+        if (revalidation.sessionId != id) {
+            // not a learner session, handle locally
+            return false;
+        }
+        itr.remove();
+        LearnerHandler learnerHandler = revalidation.handler;
+        QuorumPacket deepCopy = new QuorumPacket(qp.getType(), qp.getZxid(),
+                Arrays.copyOf(qp.getData(), qp.getData().length),
+                qp.getAuthinfo() == null ? null : new ArrayList<>(qp.getAuthinfo()));
+        learnerHandler.queuePacket(deepCopy);
+        // To keep consistent as leader, touch the session when it's
+        // revalidating the session, only update if it's a valid session.
+        if (valid) {
+            touch(revalidation.sessionId, revalidation.timeout);
+        }
+        return true;
+    }
+
+    @Override
+    public void revalidateSession(QuorumPacket qp, LearnerHandler learnerHandler) throws IOException {
+        ByteArrayInputStream bis = new ByteArrayInputStream(qp.getData());
+        DataInputStream dis = new DataInputStream(bis);
+        long id = dis.readLong();
+        int to = dis.readInt();
+        synchronized (revalidateSessionLock) {
+            pendingRevalidations.add(new Revalidation(id, to, learnerHandler));
+            Learner learner = zks.getLearner();
+            if (learner != null) {
+                learner.writePacket(qp, true);
+            }
+        }
+    }
+
+    @Override
+    public void submitLearnerRequest(Request si) {
```

Comment:
Also add some comments here to show why we add the request to the firstProcessor without going through the validation (same as leader), and show why we have it going through the whole pipeline of this ObserverMaster (I think our decision at that time is make it simpler and identical with leader).

Add those comments will show the thoughts why we're doing this, and make the review easier.